core/stdlib/std.ncl

```diff
@@ -4713,6 +4713,96 @@
+      | doc m%"
+        `base64_encode variant str` base64 encodes the value of a string using the given variant
+         of the encoding.
+
+         The encoding options are:
+         - 'Standard: Use the standard base64 alphabet and include padding.
+         - 'UrlSafe: Use the url-safe variant of the base64 alphabet, with padding.
+         - 'NoPad: Use the standard alphabet with no padding.
+         - 'UrlSafeNoPad: Use the url-safe alphabet without padding.
+
+        # Examples
+
+        ```nickel multiline
+        std.string.base64_encode 'Standard "Hello, world!"
+        # => "SGVsbG8sIHdvcmxkIQ=="
+
+        std.string.base64_encode 'NoPad "Hello, world!"
+        # => "SGVsbG8sIHdvcmxkIQ"
+
+        std.string.base64_encode 'UrlSafe "😎"
+        # => "8J-Yjg=="
+
+        std.string.base64_encode 'UrlSafeNoPad "😎"
+        # => "8J-Yjg"
+        ```
+      "%
+      = fun variant s => %string/base64_encode% variant s,
+
+    base64_decode
+      : [| 'Standard, 'UrlSafe, 'NoPad, 'UrlSafeNoPad |] -> String -> String
+      | doc m%"
+        `base64_decode variant str` decodes the value of the string using the given variant of
+         base64 encoding.
+
+         The encoding options are:
+         - 'Standard: Use the standard base64 alphabet and include padding.
+         - 'UrlSafe: Use the url-safe variant of the base64 alphabet, with padding.
+         - 'NoPad: Use the standard alphabet with no padding.
+         - 'UrlSafeNoPad: Use the url-safe alphabet without padding.
+
+        This may fail if the input string is not valid base64 encoded data. It may also fail if
+        the decoding succeeds, but the decoded value is not a valid utf-8 string.
+
+        # Examples
+
+        ```nickel multiline
+        std.string.base64_decode 'Standard "SGVsbG8sIHdvcmxkIQ=="
+        # => "Hello, world!"
+
+        std.string.base64_decode 'NoPad "SGVsbG8sIHdvcmxkIQ"
+        # => "Hello, world!"
+
+        std.string.base64_decode 'UrlSafe "8J-Yjg=="
+        # => "😎"
+
+        std.string.base64_decode 'UrlSafeNoPad "8J-Yjg"
+        # => "😎"
+        ```
+      "%
+      = fun variant str =>
+        let Base64String =
+          %contract/custom% (fun _label value =>
+            (
+              %typeof% value
+              |> match {
+                'Enum => 'Ok (value | [| 'Decoded String |]),
+                'String =>
+                  (%string/base64_decode% variant (value | String))
+                  |> match {
+                    'Ok dec => 'Ok ('Decoded dec),
+                    'Error err => 'Error err
+                  },
+                _ => 'Error { message = "Expected a String or a [| 'Decoded String |]" }
+              }
+            )
+              | [|
+                'Ok Dyn,
+                'Error {
+                  message | String | optional,
+                  notes | Array String | optional,
+                  blame_location | Dyn | optional
+                }
+              |]
```

Comment:
Personally, I find the static typing of custom contracts annoying enough that I usually just do `let Base64String | Dyn = ...` to turn it off.